tiledb/sm/c_api/tiledb.h

```diff
@@ -4497,6 +4609,861 @@ TILEDB_EXPORT int32_t tiledb_query_get_fragment_timestamp_range(
     uint64_t* t1,
     uint64_t* t2);
 
+/**
+ * return a TileDB subarray object for the given array.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_subarray_t* subarray;
+ * tiledb_query_get_subarray(array, &subarray);
+ * @endcode
+ *
+ * @param ctx The TileDB context.
+ * @param array An open array object.
+ * @param subarray The retrieved subarray object if available.
+ * @return `TILEDB_OK` for success and `TILEDB_OOM` or `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT
+int32_t tiledb_query_get_subarray(
+    tiledb_ctx_t* ctx,
+    const tiledb_query_t* query,
+    tiledb_subarray_t** subarray);
+
+/* ********************************* */
+/*             SUBARRAY              */
+/* ********************************* */
+
+/**
+ * Populates a subarray with specific indicies.
+ *
+ * **Example:**
+ *
+ * The following sets a 2D subarray [0,10], [20, 30] to the subarray.
+ *
+ * @code{.c}
+ * tiledb_subarray_t *subarray;
+ * uint64_t subarray_a[] = { 0, 10, 20, 30};
+ * tiledb_subarray_set_subarray(ctx, subarray, subarray_a);
+ * @endcode
+ *
+ * @param ctx The TileDB context.
+ * @param subarray The TileDB subarray object.
+ * @param subarray_a The subarray which can be used to constrain the array
+ read/write.
+ *     It should be a sequence of [low, high] pairs (one
+ *     pair per dimension). For the case of writes, this is meaningful only
+ *     for dense arrays, and specifically dense writes. Note that `subarray_a`
+ *     must have the same type as the domain.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ *
+ * TBD: These notes (related to query functionality) 
+ * perhaps become incorporated into those for
+ * tiledb_query_set_subarray_v2(), modified as needed???
+ x* @note If you set the subarray of a completed, incomplete or in-progress
+ x*     query, this function will clear the internal state and render it
+ x*     as uninitialized. However, the potentially set layout and attribute
+ x*     buffers will be retained. This is useful when the user wishes to
+ x*     fix the attributes and layout, but explore different subarrays with
+ x*     the same `tiledb_query_t` object (i.e., without having to create
+ x*     a new object).
+ *
+ x* @note Setting the subarray in sparse writes is meaningless and, thus,
+ x*     this function will error in the following two cases, provided that
+ x*     this is a write query:
+ x*     (i) the array is sparse, and (ii) the array is dense and the
+ x*     layout has been set to `TILEDB_UNORDERED`. In the second case,
+ x*     if the user sets the layout to `TILEDB_UNORDERED` **after**
+ x*     the subarray has been set, the subarray will simply be ignored.
+ */
+TILEDB_EXPORT
+int32_t tiledb_subarray_set_subarray(
```

Comment:
I understand the motivation for this API. Ideally, we want users to set the ranges individually. Is it feasible to remove this and refactor existing usage to set ranges instead?